File: L4/App_4_1/App_4_1.cpp
/**
 * @brief PCLP1 | L4
 *      Determinarea daca un numar este prim si perfect.
 *
 * @file App_4_1.cpp
 */

#include <iostream>

bool is_prime(int n)
{
    if (n <= 1)
        return false;

    for (unsigned int d = 2, N = n / 2; d <= N; ++d)
        if (n % d == 0)
            return false;

    return true;
}

bool is_perfect(int n)
{
    n = (n < 0) ? -n : n;

    unsigned int sum = 1;

    for (unsigned int d = 2, N = n / 2; d <= N; ++d)
        if (n % d == 0)
            sum += d;

    return sum == n;
}

int main()
{
    int n = 0;

    std::cout << "n = ";
    std::cin >> n;

    std::cout << "Prime:" << is_prime(n) << "\nPerfect: " << is_perfect(n);

    return 0;
}

File: L4/App_4_1/App_4_1.py
# 
# PCLP1 | L4
# Determinarea daca un numar este prim si perfect.
# 
# File: App_4_1.py
# 

def is_prime(n):
    if n <= 1:
        return False

    for d in range(2, int(n / 2) + 1):
        if n % d == 0:
            return False
    
    return True

def is_perfect(n):
    n = abs(n)
    
    sum = 1

    for d in range(2, int(n / 2) + 1):
        if n % d == 0:
            sum = sum + d

    return sum == n

def main():
    n = int(input("n = "))

    print(f"Prime: {is_prime(n)}")
    print(f"Perfect: {is_perfect(n)}")

if __name__ == "__main__":
    main()  